### PR TITLE
[DFT] add backend_selector to compile-time dispatch compute_xxxward

### LINKED ISSUE #309: [DFT] Building multiple DFT backends at once causes linking conflicts
# Summary
When building multiple DFT backends at once, each backend defines a symbol for the many `oneapi::mkl::dft::compute_forward(...)` and `oneapi::mkl::dft::compute_backward(...)` compilation time dispatch overloads, so at linking time either one can be chosen.
To get around this I tried linking just one of the two backends, but then compilation failed because of an undefined reference to `oneapi::mkl::dft::detail::descriptor<...>::commit(oneapi::mkl::backend_selector<other_backend>)`. This is because both `ENABLE_MKLCPU_BACKEND`  and `ENABLE_MKLGPU_BACKEND` are defined in the preprocessor so both commits are defined here - 
https://github.com/oneapi-src/oneMKL/blob/develop/include/oneapi/mkl/dft/detail/descriptor_impl.hpp#L63 .
I tried getting around this by providing a definition of `commit` but made it a weak symbol so it can be overridden when the appropriate backend is linked. This led to problems when running the tests because https://github.com/oneapi-src/oneMKL/blob/develop/tests/unit_tests/main_test.cpp#L113 is still picking up both devices.

I believe other interfaces get around this by having a `backend_selector<backend::BACKEND>` parameter on all the functions, like here https://github.com/oneapi-src/oneMKL/blob/develop/include/oneapi/mkl/blas/detail/blas_ct_backends.hxx. I think the best course of action might be a spec change to include a `backend_selector` on `compute_forward` and `compute_backward`.

# Version
This on develop, I specifically tested githash 005c3c79a44974fc8c3adebfa71d94b0fe005f33

# Steps to reproduce
Build with both ENABLE_MKLCPU_BACKEND and ENABLE_MKLGPU_BACKEND set to 1.

# Observed behavior
When running `test_main_dft_ct` I get this output:
```
finlay@devpc-c310:~/perf/oneMKL/build$ ./bin/test_main_dft_ct 
Run this program with --terse_output to change the way it prints its output.
[==========] Running 488 tests from 3 test suites.
[----------] Global test environment set-up.
[----------] 240 tests from ComputeTestSuite/ComputeTests_in_place
[ RUN      ] ComputeTestSuite/ComputeTests_in_place.COMPLEX_SINGLE_in_place_buffer/sizes_8_batches_1_Intel_R__UHD_Graphics_750__0x4c8a_
Skipping test because: "oneMKL: DFT/compute_forward: function is not implemented Not implemented for MKLCPU"
...
```
# Expected behavior
The MKLGPU `compute_forward` should be called.

### PR DESCRIPTION
**Pull request by FMarno**
# Description
Add a `backend_selector` argument to all the backend specific `compute_xxxward` functions. This is only there to give the function a unique signature, to avoid linking conflicts.

Fixes #309

# Checklist

## All Submissions

- [x] Do all unit tests pass locally? Attach a log.
- [x] Have you formatted the code using clang-format?

## New interfaces

- [ ] Have you provided motivation for adding a new feature as part of RFC and
it was accepted? # (RFC)
- [ ] What version of oneAPI spec the interface is targeted?
- [ ] Complete [New features](pull_request_template.md#new-features) checklist

## New features

- [x] Have you provided motivation for adding a new feature?
- [x] Have you added relevant tests?

## Bug fixes

- [x] Have you added relevant regression tests?
- [x] Have you included information on how to reproduce the issue (either in a
      GitHub issue or in this PR)?

### CODE REVIEW

File: examples/dft/compile_time_dispatching/CMakeLists.txt [hunk 1]
```diff
@@ -18,46 +18,30 @@
...
-      PUBLIC ${ONEMKL_LIBRARIES_${domain}}
-      ${ONEMKL_LIBRARIES_${domain}}
-      ONEMKL::SYCL::SYCL
-  )
   # Register example as ctest
- add_test(NAME ${domain}/EXAMPLE/CT/${dfti_ct_sources} COMMAND example_${domain}_${dfti_ct_sources})
-endforeach(dfti_ct_sources)
\ No newline at end of file
+  add_test(NAME dft/EXAMPLE/CT/complex_fwd_buffer_${dfti_backend} COMMAND ${EXAMPLE_NAME})
+endforeach(dfti_backend)
```
Comment (Rbiessy) [inline examples/dft/compile_time_dispatching/CMakeLists.txt]: Nitpicking, could add a new line.

**Comment by FMarno** [comment]
test run - [mklcpu_mklgpu_test_run.log](https://github.com/oneapi-src/oneMKL/files/11569452/mklcpu_mklgpu_test_run.log)

File: tests/unit_tests/dft/source/descriptor_tests.cpp [hunk 2]
```diff
@@ -431,8 +431,7 @@ inline void recommit_values(sycl::queue& sycl_queue) {
     std::vector<test_params> argument_groups{
         // not changeable
         // FORWARD_DOMAIN, PRECISION, DIMENSION, COMMIT_STATUS
-        { std::make_pair(config_param::NUMBER_OF_TRANSFORMS, std::int64_t{ 5 }),
```
Comment (Rbiessy) [inline tests/unit_tests/dft/source/descriptor_tests.cpp:434]: Why is this needed?

File: tests/unit_tests/dft/include/test_common.hpp [hunk 3]
```diff
@@ -162,6 +162,17 @@ void commit_descriptor(oneapi::mkl::dft::descriptor<precision, domain> &descript
...
 }
 
+template <typename Func, typename... Args>
+void dispatch(sycl::queue queue, Func func, Args &&... args) {
+#ifdef CALL_RT_API
+    (void)queue;
+    func(std::forward<Args>(args)...);
+#else
+    // can't get return from this, so func must capture return values by reference in a lambda
+    TEST_RUN_CT_SELECT(queue, func, std::forward<Args>(args)...);
```
Comment (Rbiessy) [inline tests/unit_tests/dft/include/test_common.hpp:172]: Is this working for the dft domain? We've had to introduce `TEST_RUN_CT_SELECT_NO_ARGS` to use cufft for instance.

File: src/dft/backends/cufft/CMakeLists.txt [hunk 4]
```diff
@@ -48,7 +48,7 @@ target_link_libraries(${LIB_OBJ} PUBLIC ONEMKL::SYCL::SYCL ${MKL_LINK_SYCL})
 set_target_properties(${LIB_OBJ} PROPERTIES
   POSITION_INDEPENDENT_CODE ON
 )
-target_link_libraries(${LIB_NAME} PUBLIC ${LIB_OBJ})
+target_link_libraries(${LIB_NAME} PUBLIC ${LIB_OBJ} onemkl_dft)
```
Comment (Rbiessy) [inline src/dft/backends/cufft/CMakeLists.txt:51]: It makes sense to me that we add `onemkl_dft` here but it does not seem to match with other domains. I'm not sure where else `onemkl_dft` was linked before.

File: tests/unit_tests/CMakeLists.txt [hunk 5]
```diff
@@ -182,6 +182,9 @@ foreach(domain ${TARGET_DOMAINS})
       PROPERTIES TEST_PREFIX ${DOMAIN_PREFIX}/RT/
       DISCOVERY_TIMEOUT 30
     )
+
+    # ensure rt tests have up-to-date libs
+    add_dependencies(test_main_${domain}_rt ${ONEMKL_LIBRARIES_${domain}})
```
Comment (Rbiessy) [inline tests/unit_tests/CMakeLists.txt:187]: In most cases `add_dependencies` shouldn't be needed if we use `target_link_libraries`. I can see the compile tests above use `add_dependencies` but I'm not sure why, do you know if there is a good reason for it?

File: src/dft/backends/cufft/CMakeLists.txt (hunk 4 above)
Comment (FMarno) [reply src/dft/backends/cufft/CMakeLists.txt:51]: `onemkl_dft` couldn't be linked before because linking with both `onemkl_dft` and with `onemkl_dft_xxxxx` would cause multiple definitions of the `compute_xxxward` functions.
WRT the other domains, I'm not sure how that works. I'm adding `onemkl_dft` here so users get everything by just linking `onemkl_dft_xxxxx`, maintaining the current functionality. Personally I think users should specify both libraries when linking since there will potentially be multiple definitions for some functions e.g. if you tried to link both `onemkl_dft_mklcpu` and `onemkl_dft_mklgpu`.

File: tests/unit_tests/dft/source/descriptor_tests.cpp (hunk 2 above)
Comment (FMarno) [reply tests/unit_tests/dft/source/descriptor_tests.cpp:434]: mklcpu considers it an invalid configuration if `NUMBER_OF_TRANSFORMS != 1 && XWD_DISTANCE < N`

File: tests/unit_tests/dft/include/test_common.hpp (hunk 3 above)
Comment (FMarno) [reply tests/unit_tests/dft/include/test_common.hpp:172]: I added the `TEST_RUN_CT_SELECT` previously, so it works fine https://github.com/oneapi-src/oneMKL/blob/055b3701c0b1eeb7a3c33866cd9239f0ddb1015c/tests/unit_tests/include/test_helper.hpp#L175
I can post a cufft log as well

File: tests/unit_tests/CMakeLists.txt (hunk 5 above)
Comment (FMarno) [reply tests/unit_tests/CMakeLists.txt:187]: I'm not sure why they have done that for the compile-time tests. 
For this change I had a few instances were I would update the library and then try to build only the test and I would be confused why nothing changed! Is there a better way I should be doing this?

File: examples/dft/compile_time_dispatching/CMakeLists.txt (hunk 1 above)
Comment (FMarno) [reply examples/dft/compile_time_dispatching/CMakeLists.txt]: addressed in https://github.com/oneapi-src/oneMKL/pull/329/commits/80e5213fe058f8ef84710d1467996f2a58459ff9

File: tests/unit_tests/dft/include/test_common.hpp (hunk 3 above)
Comment (Rbiessy) [reply tests/unit_tests/dft/include/test_common.hpp:172]: Thanks, my bad I missed it.

**Comment by FMarno** [comment]
[cufft_run.log](https://github.com/oneapi-src/oneMKL/files/11626502/cufft_run.log)

File: tests/unit_tests/CMakeLists.txt (hunk 5 above)
Comment (Rbiessy) [reply tests/unit_tests/CMakeLists.txt:187]: I am surprised that `add_dependencies` is needed as I would expect `test_main_${domain}_rt` is already linked with `${ONEMKL_LIBRARIES_${domain}}` somehow so I am wondering if we are missing some dependencies elsewhere for the other domains.

**Comment by Rbiessy** [comment]
@mkrainiuk could you or someone else help with the review? I have left 2 comments open as I think there are some things we don't understand in the CMake today. Hopefully someone can help with that.

**Comment by FMarno** [comment]
I've converted this to a draft since we aren't sure that this is the direction we would like to take.

**Comment by FMarno** [comment]
problem was fixed with #342 instead
